jumanji/environments/logic/graph_coloring/env.py

```diff
@@ -0,0 +1,310 @@
+from jumanji.environments.logic.graph_coloring.viewer import GraphColoringViewer
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class GraphColoring(Environment[State]):
+    """Environment for the GraphColoring problem.
+    The problem is a combinatorial optimization task where the goal is
+      to assign a color to each vertex of a graph
+      in such a way that no two adjacent vertices share the same color.
+    The problem is usually formulated as minimizing the number of colors used.
+
+    - observation: `Observation`
+        - adj_matrix: jax array (bool) of shape (num_nodes, num_nodes),
+            representing the adjacency matrix of the graph.
+        - colors: jax array (int32) of shape (num_nodes,),
+            representing the current color assignments for the vertices.
+        - action_mask: jax array (bool) of shape (num_colors,),
+            indicating which actions are valid in the current state of the environment.
+        - current_node_index: integer representing the current node being colored.
+
+    - action: int, the color to be assigned to the current node (0 to num_nodes - 1)
+
+     - reward: float, a sparse reward is provided at the end of the episode.
+        Equals the negative of the number of unique colors used to color all vertices in the graph.
+        If an invalid action is taken, the reward is the negative of the total number of colors.
+
+    - episode termination:
+        - if all nodes have been assigned a color or if an invalid action is taken.
+
+    - state: `State`
+        - adj_matrix: jax array (bool) of shape (num_nodes, num_nodes),
+            representing the adjacency matrix of the graph.
+        - colors: jax array (int32) of shape (num_nodes,),
+            color assigned to each node, -1 if not assigned.
+        - current_node_index: jax array (int) with shape (),
+            index of the current node.
+        - action_mask: jax array (bool) of shape (num_colors,),
+            indicating which actions are valid in the current state of the environment.
+        - key: jax array (uint32) of shape (2,),
+            random key used to generate random numbers at each step and for auto-reset.
+
+    ```python
+    from jumanji.environments import GraphColoring
+    env = GraphColoring()
+    key = jax.random.key(0)
+    state, timestep = jax.jit(env.reset)(key)
+    env.render(state)
+    action = env.action_spec().generate_value()
+    state, timestep = jax.jit(env.step)(state, action)
+    env.render(state)
+    ```
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        """Instantiate a `GraphColoring` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator`.
```

Comment:
Is it worth specifying which parameters this random generator is instantiated with? I.e. 100 nodes where 80% of the nodes are connected? I think other environments have done that, but do correct me if I'm wrong

Reply:
The Cleaner and Connector environments specified the default values for example :)

Reply:
we have that right on the 95th line though 😄

Reply:
Ah my bad, I'd totally missed that! All good, will resolve this comment 😄